### PR TITLE
Add gnp_random_bqm function for generating random BQMs

### PR DESCRIPTION
**Pull request by arcondello**
Generator for random BQMs. Will be performant once https://github.com/dwavesystems/dimod/issues/653 is implemented.

### CODE REVIEW

**Review by JoelPasvolsky — APPROVED** [review]

File: dimod/generators/random.py [hunk 1]
```diff
@@ -11,17 +11,93 @@
...
+
+        random_state (:class:`numpy.random.RandomState`/int, optional):
+            A random seed or a random state generator. Used for generating
+            the structure of the BQM and, if `data_rvs` is not given, for
+            the bias generation.
+
+        bias_generator (callable, optional):
+            Bias generating function.
+            Should accept a single argument specifying the length of the
+            ndarray that it will return.
```
[the remark this answers is not in the document]
Comment (arcondello) [reply dimod/generators/random.py:48]: This parameter is not intended for high-level users. Further, it happens in this case to be handed the number of biases but that is not in general true.

File: dimod/generators/random.py [hunk 2]
```diff
@@ -11,17 +11,93 @@
...
+    Args:
+        n (int/sequence):
+            The variables. If `n` is an int, the variables are labelled as
+            `[0, n)`.
+
+        p (float): The probability for interaction creation.
+
+        cls (:class:`.BinaryQuadraticModel`): BQM class to build from.
+
+        random_state (:class:`numpy.random.RandomState`/int, optional):
```
[the remark this answers is not in the document]
Comment (arcondello) [reply dimod/generators/random.py:40]: Different behavior. `seed` in the others is a number, here it is a `RandomState`.

File: dimod/generators/random.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (arcondello) [reply dimod/generators/random.py:48]: Reworked the language a little bit.

**Comment by arcondello** [comment]
Blocked for the moment by https://github.com/dwavesystems/dimod/issues/740. We could just change the test to let is pass, but best to fix the bug first.

File: dimod/generators/random.py [hunk 3]
```diff
@@ -11,17 +11,98 @@
...
+    if isinstance(n, abc.Sequence):
+        labels = n
+        n = len(labels)
+    else:
+        labels = None
+
+    if not isinstance(random_state, np.random.RandomState):
+        random_state = np.random.RandomState(random_state)
+
+    neighbborhoods = []
```
Comment (hhtong) [inline dimod/generators/random.py:75]: Typo

File: dimod/generators/random.py [hunk 4]
```diff
@@ -11,17 +11,98 @@
...
+
+        p (float): The probability for interaction creation.
+
+        vartype (:class:`.Vartype`/str/set):
+            Variable type for the BQM. Accepted input values:
+
+            * :class:`.Vartype.SPIN`, ``'SPIN'``, ``{-1, 1}``
+            * :class:`.Vartype.BINARY`, ``'BINARY'``, ``{0, 1}``
+
+        cls (:class:`.BinaryQuadraticModel`): BQM class to build from.
```
Comment (hhtong) [inline dimod/generators/random.py:44]: Missing: 'optional' ?

File: dimod/generators/random.py [hunk 5]
```diff
@@ -11,17 +11,98 @@
...
+        q += len(neighbors)
+
+    # calculate the biases
+    if bias_generator is None:
+        def bias_generator(n):
+            return random_state.uniform(size=n)
+
+    ldata = bias_generator(n)
+    qdata = bias_generator(num_interactions)
+    offset, = bias_generator(1)
```
Comment (hhtong) [inline dimod/generators/random.py:102]: Extra comma

**Review by hhtong — APPROVED** [review]

**Pull request merged**